src/decoder/mod.rs

```diff
@@ -1066,7 +1066,10 @@ impl<R: Read + Seek> Decoder<R> {
         let max_sample_bits = self.image().bits_per_sample;
         match self.image().sample_format {
             SampleFormat::Uint => match max_sample_bits {
-                n if n <= 8 => DecodingResult::new_u8(buffer_size, &self.limits),
+                n if n < 8 => {
+                    DecodingResult::new_u8(buffer_size / 8 * usize::from(n), &self.limits)
```

Comment:
I'm slightly suspicious that this will round incorrectly if the buffer_size isn't a multiple of 8